ament_pclint/ament_pclint/main.py

```diff
@@ -0,0 +1,372 @@
+
+    aggregate_xml = _format_pclint_xml_out(result)
+
+    try:
+        root = ElementTree.fromstring(aggregate_xml)
+    except ElementTree.ParseError as e:
+        print('Invalid XML in pclint output: %s' % str(e),
+              file=sys.stderr)
+        return 1
+
+    # output errors
+    report = {}
+    for doc in root.findall('doc'):
+        for error in doc.findall('error'):
+            report[error.find('file').text] = []
+        for error in doc.findall('error'):
+            filename = error.find('file').text
+            data = {
+                'line': int(error.find('line').text),
+                'id': error.find('code').text,
+                'severity': error.find('type').text,
+                'msg': error.find('desc').text,
+            }
+            report[filename].append(data)
+
+            data = dict(data)
+            data['filename'] = filename
+            print('[%(filename)s:%(line)d]: (%(severity)s: %(id)s) %(msg)s' % data,
+                  file=sys.stderr)
+
+    # output summary
+    error_count = sum(len(r) for r in report.values())
+    error_count += sum(retvals)
+    if not error_count:
+        print('No errors')
+        rc = 0
+    else:
+        print('%d errors' % error_count, file=sys.stderr)
+        if args.debug:
+            for r in result:
+                print(r)
+        rc = 1
+
+    # generate xunit file
+    if args.xunit_file:
+        folder_name = os.path.basename(os.path.dirname(args.xunit_file))
+        file_name = os.path.basename(args.xunit_file)
+        suffix = '.xml'
+        if file_name.endswith(suffix):
+            file_name = file_name[0:-len(suffix)]
+            suffix = '.xunit'
+            if file_name.endswith(suffix):
+                file_name = file_name[0:-len(suffix)]
+        testname = '%s.%s' % (folder_name, file_name)
+
+        xml = get_xunit_content(report, testname, time.time() - start_time)
+        path = os.path.dirname(os.path.abspath(args.xunit_file))
+        if not os.path.exists(path):
+            os.makedirs(path)
+        with open(args.xunit_file, 'w') as f:
+            f.write(xml)
+
+    return rc
+
+
+def execute_analysis(cmd, result, retvals, idx):
+    try:
+        p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        xml = p.communicate()[0].decode("utf-8")  # Read xml from stdout
+    except subprocess.CalledProcessError as e:
+        print("The invocation of 'pclint' failed with error code %d: %s" %
+              (e.returncode, e), file=sys.stderr)
+        return False
+    result[idx] = xml
+    retvals[idx] += p.returncode
+
+
+def find_executable(file_name, additional_paths=None):
+    if sys.platform == 'linux':
+        file_name = file_name + '_linux'
+    elif sys.platform == 'darwin':
+        file_name = file_name + '_osx'
+    path = None
+    if additional_paths:
+        path = os.getenv('PATH', os.defpath)
+        path += os.path.pathsep + os.path.pathsep.join(additional_paths)
```

Comment:
`additional_paths` doesn't seem to be used anywhere. Should it just be removed? I was mostly wondering if appending of additional paths is the  "right" way...

Reply:
I was just following what `ament_cppcheck` is doing: https://github.com/ament/ament_lint/blob/master/ament_cppcheck/ament_cppcheck/main.py#L158-L163

But, it's up to you, I can delete it.

Reply:
I am not saying that it should be removed. I just want to point out that it might not work for packages which use target specific properties (which is the modern way in CMake).